use_cases/eluc/prescriptors/nsga2/torch_prescriptor.py

```diff
@@ -0,0 +1,265 @@
+                                      np.zeros((len(encoded_context_df), len(constants.RECO_COLS))))
+            context_dl = torch.utils.data.DataLoader(context_ds, batch_size=self.batch_size, shuffle=False)
+        else:
+            context_df = self.eval_df
+            context_dl = self.context_dl
+
+        # Aggregate recommendations
+        reco_list = []
+        with torch.no_grad():
+            for X, _ in context_dl:
+                recos = candidate(X)
+                reco_list.append(recos)
+            reco_tensor = torch.concatenate(reco_list, dim=0)
+
+            # Convert recommendations into context + actions
+            reco_df = self._reco_tensor_to_df(reco_tensor, context_df)
+
+        context_actions_df = self._reco_to_context_actions(reco_df, context_df)
+        return context_actions_df
+
+    def predict_metrics(self, context_actions_df: pd.DataFrame) -> pd.DataFrame:
+        """
+        Computes ELUC and change for each sample in a context_actions_df.
+        """
+        eluc_df = self.predictor.predict(context_actions_df)
+        change_df = self._compute_percent_changed(context_actions_df)
+        
+        return eluc_df, change_df
+
+    def _evaluate_candidates(self, candidates: list[Candidate]):
+        """
+        Calls prescribe and predict on candidates and assigns their metrics to the results.
+        """
+        for candidate in candidates:
+            context_actions_df = self._prescribe(candidate)
+            eluc_df, change_df = self.predict_metrics(context_actions_df)
+            candidate.metrics = (eluc_df["ELUC"].mean(), change_df["change"].mean())
+
+    def _select_parents(self, candidates: list[Candidate], n_parents: int) -> list[Candidate]:
+        """
+        NSGA-II parent selection using fast non-dominated sort and crowding distance.
+        Sets candidates' ranks and distance attributes.
+        """
+        fronts, ranks = nsga2_utils.fast_non_dominated_sort(candidates)
+        for candidate, rank in zip(candidates, ranks):
+            candidate.rank = rank
+        parents = []
+        for front in fronts:
+            # Compute crowding distance here even though it's technically not necessary now
+            # so that later we can sort by distance
+            crowding_distance = nsga2_utils.calculate_crowding_distance(front)
+            for candidate, distance in zip(front, crowding_distance):
+                candidate.distance = distance
+            if len(parents) + len(front) > n_parents:  # If adding this front exceeds num_parents
+                front = sorted(front, key=lambda c: c.distance, reverse=True)
+                parents += front[:n_parents - len(parents)]
+                break
+            parents += front
+        return parents
+    
+    def _tournament_selection(self, sorted_parents: list[Candidate]) -> tuple[Candidate, Candidate]:
+        """
+        Same implementation as in ESP.
```

Comment:
ESP is not open source, at least for now. I think you can remove references to it.